### PR TITLE
ARROW-16686: [C++] Use shared_ptr with FunctionOptions

### PR DESCRIPTION
**Pull request by vibhatha**
Includes changes to remove raw pointer usage with `FunctionOptions` used in `Aggregate`.

This also includes changes in removing `keep_alives` from R aggregate bindings as well.

### CODE REVIEW

**Comment by vibhatha** [comment]
cc @westonpace should the owned smart pointers be removed?

**Comment by vibhatha** [comment]
@kou need help with the C GLib changes.

**Comment by vibhatha** [comment]
cc @nealrichardson made some changes to R library, appreciate your feedback.

**Comment by kou** [comment]
Done.

**Comment by vibhatha** [comment]
> Done.

Thank you!

**Comment by nealrichardson** [comment]
> cc @nealrichardson made some changes to R library, appreciate your feedback.

Tests pass so LGTM

**Comment by vibhatha** [comment]
@kou After rebase the C Glib CI is broken. Could you please help?

**Comment by kou** [comment]
I"ve reapplied my patches.

**Comment by vibhatha** [comment]
> I've reapplied my patches.

did my rebase caused any issue?

**Comment by kou** [comment]
I think that you got conflicts when you run `git rebase apache/master`. I think that you run `git rebase --skip` for these conflicts. It dropped my commits.

**Comment by vibhatha** [comment]
> I think that you got conflicts when you run `git rebase apache/master`. I think that you run `git rebase --skip` for these conflicts. It dropped my commits.

I ran `git rebase --continue` after adding the modified files. Sorry about that.

**Comment by vibhatha** [comment]
> I think that you got conflicts when you run `git rebase apache/master`. I think that you run `git rebase --skip` for these conflicts. It dropped my commits.

I ran `git rebase --continue` after adding the modified files. Sorry about that.

**Comment by vibhatha** [comment]
cc @lidavidm @westonpace could you please take a look.

File: cpp/src/arrow/compute/api_aggregate.h [hunk 1]
```diff
@@ -401,7 +401,7 @@ struct ARROW_EXPORT Aggregate {
   std::string function;
 
   /// options for the aggregation function
-  const FunctionOptions* options;
+  std::shared_ptr<FunctionOptions> options;
```
Comment (lidavidm) [inline cpp/src/arrow/compute/api_aggregate.h:404]: is `unique_ptr` possible?

File: cpp/src/arrow/compute/exec/aggregate.cc [hunk 2]
```diff
@@ -61,7 +61,11 @@ Result<std::vector<std::unique_ptr<KernelState>>> InitKernels(
       // use known default options for the named function if possible
       auto maybe_function = ctx->func_registry()->GetFunction(aggregates[i].function);
       if (maybe_function.ok()) {
-        options = maybe_function.ValueOrDie()->default_options();
+        auto default_opts = maybe_function.ValueOrDie()->default_options();
+        // only update the options if the default options are set
+        if (default_opts != nullptr) {
+          options = default_opts->Copy();
+        }
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/aggregate.cc:68]:
```suggestion
        FunctionOptions* default_opts = maybe_function.ValueOrDie()->default_options();
        if (default_opts != nullptr) {
          options = default_opts->Copy();
        }
```

File: cpp/src/arrow/compute/exec/aggregate_node.cc [hunk 3]
```diff
@@ -116,11 +111,8 @@ class ScalarAggregateNode : public ExecNode {
       kernels[i] = static_cast<const ScalarAggregateKernel*>(kernel);
 
       if (aggregates[i].options == nullptr) {
-        aggregates[i].options = function->default_options();
-      }
-      if (aggregates[i].options) {
-        owned_options[i] = aggregates[i].options->Copy();
-        aggregates[i].options = owned_options[i].get();
+        auto opts = function->default_options()->Copy();
+        aggregates[i].options = std::move(opts);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/aggregate_node.cc:115]:
```suggestion
        aggregates[i].options = function->default_options()->Copy();
```

File: python/pyarrow/_compute.pyx [hunk 4]
```diff
@@ -2067,16 +2067,18 @@ def _group_by(args, keys, aggregations):
...
+        shared_ptr[CFunctionOptions] null_opts
 
     _pack_compute_args(args, &c_args)
     _pack_compute_args(keys, &c_keys)
 
     for aggr_func_name, aggr_opts in aggregations:
         c_aggr.function = tobytes(aggr_func_name)
         if aggr_opts is not None:
-            c_aggr.options = (<FunctionOptions?> aggr_opts).get_options()
+            # (<FunctionOptions?> aggr_opts).get_options()
```
Comment (lidavidm) [inline python/pyarrow/_compute.pyx:2078]: what is the comment for?

File: python/pyarrow/_compute.pyx [hunk 5]
```diff
@@ -2067,16 +2067,18 @@ def _group_by(args, keys, aggregations):
...
+            c_aggr.options = (<FunctionOptions?>aggr_opts).wrapped
         else:
-            c_aggr.options = NULL
+            c_aggr.options = null_opts
```
Comment (lidavidm) [inline python/pyarrow/_compute.pyx:2081]: `<shared_ptr[CFunctionOptions]>nullptr` should work

File: cpp/src/arrow/compute/api_aggregate.h (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/compute/api_aggregate.h:404]: Good point, initially the idea was to put a `unique_ptr`, but there was a discussion in the JIRA and finally settled for the `shared_ptr`.

File: cpp/src/arrow/compute/exec/aggregate.cc (hunk 2 above)
Comment (vibhatha) [reply cpp/src/arrow/compute/exec/aggregate.cc:68]: Good one, I forgot the recent discussion on auto.

File: cpp/src/arrow/compute/api_aggregate.h (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/api_aggregate.h:404]: Ah, I didn't see that.

Hmm, I suppose it's not much of a difference here.

File: python/pyarrow/_compute.pyx (hunk 4 above)
Comment (vibhatha) [reply python/pyarrow/_compute.pyx:2078]: Ah that's typo. I will remove this.

File: cpp/src/arrow/compute/exec/aggregate.cc [hunk 6]
```diff
@@ -61,7 +61,10 @@ Result<std::vector<std::unique_ptr<KernelState>>> InitKernels(
       // use known default options for the named function if possible
       auto maybe_function = ctx->func_registry()->GetFunction(aggregates[i].function);
       if (maybe_function.ok()) {
-        options = maybe_function.ValueOrDie()->default_options();
+        FunctionOptions* default_opts = maybe_function.ValueOrDie()->default_options();
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/aggregate.cc:64]: Hmm, I guess this doesn't work. Maybe `std::move(maybe_function).MoveValueUnsafe()->default_options()` or something?

File: cpp/src/arrow/compute/exec/aggregate.cc [hunk 7]
```diff
@@ -61,7 +61,11 @@ Result<std::vector<std::unique_ptr<KernelState>>> InitKernels(
       // use known default options for the named function if possible
       auto maybe_function = ctx->func_registry()->GetFunction(aggregates[i].function);
       if (maybe_function.ok()) {
-        options = maybe_function.ValueOrDie()->default_options();
+        const FunctionOptions* default_opts =
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/aggregate.cc:64]: I think if you change line 58 to...

```
const FunctionOptions* options = aggregates[i].options.get();
```

...then this change isn't needed.  We are just grabbing a read-only view of the options which we can safely discard after the call to `init`

File: cpp/src/arrow/compute/exec/plan_test.cc [hunk 8]
```diff
@@ -390,7 +391,8 @@ TEST(ExecPlan, ToString) {
                           }}},
               {"aggregate",
                AggregateNodeOptions{
-                   /*aggregates=*/{{"hash_sum", nullptr}, {"hash_count", &options}},
+                   /*aggregates=*/{{"hash_sum", nullptr},
+                                   {"hash_count", std::move(options)}},
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/plan_test.cc:395]: You use `options` again on line 437 so I'm not sure it is safe to move it here.

**Review by westonpace — COMMENTED** [review]
Looks good.  One minor suggestion.  Thanks for taking this on.  Looks like it ended up being considerably more work than I had expected.

File: cpp/src/arrow/compute/exec/aggregate.cc (hunk 7 above)
Comment (vibhatha) [reply cpp/src/arrow/compute/exec/aggregate.cc:64]: Good catch. Thank you.

**Comment by vibhatha** [comment]
@westonpace updated the PR.

**Review by lidavidm — APPROVED** [review]

**Review by westonpace — APPROVED** [review]

**Pull request merged**